### PR TITLE
ui: replace yarn with pnpm and reintroduce workspace support

### PR DESCRIPTION
**Pull request by sjbarag**
Bazel's rules_js requires a pnpm-lock.yaml file --- the dependency
lockfile format for the pnpm package manager [1] --- to be committed.
While it offers to translate yarn.lock files at build-time, the presence
of multiple package managers' lockfiles is slightly confusing and
increases noise in commit logs, pull requests, etc. Unlike the older
rules_nodejs, rules_js *does* support package-manager workspaces,
allowing for a single 'pnpm install' outside of bazel to install all
dependencies in pkg/ui/ (like we did pre-rules_nodejs) and deduplicated
dependency downloads across JS packages. Remove yarn and replace it with
pnpm, since that's ultimately what rules_js mimicks.

For posterity, this is the current timeline of JS package management
in CRDB:

* Jun 2015 [2]: npm added for JS package management
* Oct 2016 [3]: npm replaced with yarn
* Apr 2021 [4]: Bazel's rules_nodejs 3.7 configured
* Jun 2021 [5]: yarn workspaces support introduced
* Sep 2021 [6]: Bazel's rules_nodejs upgraded to 4.1
* May 2022 [7]: Bazel's rules_nodejs upgraded to 4.5 and yarn workspaces
                support removed
* May 2023 [8]: Bazel's rules_nodejs replaced with rules_js
* Jul 2023 [9]: yarn replaced with pnpm + pnpm workspaces

[1] https://pnpm.io/
[2] https://github.com/cockroachdb/cockroach/commit/4379fc1f6711051b339390b85bc15e8de96f31b0 (Add node package management to the ui., 2015-06-29)
[3] https://github.com/cockroachdb/cockroach/commit/8071dc50d91f9e089739aa63eb44665962a36545 (*: switch from npm to yarn, 2016-10-21)
[4] https://github.com/cockroachdb/cockroach/commit/6cc3af3a42c13df7a1f23d9a5f51b25fa8b00619 (ui: bazel build for UI, 2021-04-22)
[5] https://github.com/cockroachdb/cockroach/commit/b8323204a3e13d0ff49b2234f6fa2b8907f5fb4e (ui: use yarn workspaces, 2021-06-28)
[6] https://github.com/cockroachdb/cockroach/commit/c4594b82fa1b5e28bca94ef1ebb2eaf60c81b939 (update build_bazel_rules_nodejs build rules to version 4.1.0, 2021-09-16)
[7] https://github.com/cockroachdb/cockroach/commit/5b6c271b1cd78323ac11cde677834d4918c183f4 (bazel: upgrade to rules_nodejs 5.4.2, 2022-05-17)
[8] https://github.com/cockroachdb/cockroach/commit/3d95f5fa79de8f984b48541e6e0e5570160f5492 (build,bazel: upgrade to `rules_js`, 2023-05-18)
[9] This commit

Release note: None
Epic: None

### CODE REVIEW

**Review by sjbarag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106720)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rickystewart and @sjbarag)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_AoZ7cBq-39dTlELNl):*
I'd set this up while debugging so I could include JS source maps in the compiled `cockroach` binary. Seemed handy and it works well, so it felt worth keeping around.
___
*[`WORKSPACE` line 251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_Aormg2ojMqf2m_gdT:-N_Aormg2ojMqf2m_gdU:b5ibv9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/WORKSPACE#L251)):*
> ```Starlark
>     url = "https://storage.googleapis.com/cockroach-npm-deps/pnpm/-/pnpm-8.5.1.tgz",
>     version = "8.5.1",
>     # Declare an @pnpm//:pnpm rule that can be called externally.
> ```

This isn't actually being used. I set it up so we _could_ use it in `dev` though, and so anyone without a local `pnpm` installation can get unblocked if they need. Seems handy and low-risk to keep around.
___
*[`WORKSPACE` line 276 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_Ap2iF4eN4KM_ke1-X:-N_Ap2iF4eN4KM_ke1-Y:b-gmzkfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/WORKSPACE#L276)):*
> ```Starlark
>         "//pkg/ui/workspaces/e2e-tests:package.json",
>     ],
>     npmrc = "//pkg/ui:.npmrc.bazel",
> ```

Now that everything uses pnpm, `.npmrc.bazel` felt like a more fitting name.
___
*[`.aspect/rules/external_repository_action_cache/npm_translate_lock_LTE1NzE0OTU1MTU=` line 0 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_Ap7LIEOPZ9So1S-7Z:-N_Ap7LIEOPZ9So1S-7_:b8x4jbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/.aspect/rules/external_repository_action_cache/npm_translate_lock_LTE1NzE0OTU1MTU%3D#L0)):*
Now that we're no longer setting `update_pnpm_lock = True`, these files aren't necessary!
___
*[`build/bootstrap/bootstrap-debian.sh` line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_ApL7Z-8SjvsE5TEJq:-N_ApL7Z-8SjvsE5TEJr:bgdvbmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/build/bootstrap/bootstrap-debian.sh#L27)):*
> ```Shell
> 
> # pnpm doesn't provide a Debian repository, and supports either `curl | sh` or `npm install -g` installations.
> curl -fsSL https://get.pnpm.io/install.sh | env PNPM_VERSION=8.6.6 sh -
> ```

@ me if you prefer `npm install -g`. Neither option is great.
___
*[`pkg/cmd/generate-distdir/main.go` line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_ApUol9Ug61S54BMY2:-N_ApUol9Ug61S54BMY3:b-uf6cgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/269b9e31c4befe53181340503663ae6e3e8ed07e/pkg/cmd/generate-distdir/main.go#L166)):*
> ```Go
> }
> 
> func getShasFromYarnRepositoriesCall(call *syntax.CallExpr, shas map[string]string) error {
> ```

This didn't seem to be used before this commit. I removed it while I was in the neighborhood, but let me know if I've missed something.
___
*[`pkg/cmd/mirror/npm/main.go` line 33 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_ApdboBHFnhEpziJzD:-N_ApdboBHFnhEpziJzE:bgjxj4p) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/cmd/mirror/npm/main.go#L33)):*
> ```Go
> )
> 
> // getUnmirroredUrls finds packages in each provided pnpm-lock.yaml (each of
> ```

The entire mirroring process has been simplified! Now we only:

1. List all files in our GCS bucket
2. Create a list of `.tgz` file paths that pnpm would download from the public registry, based on `pnpm-lock.yaml`
3. Compare the paths in both lists
4. Download/upload only the packages that aren't already mirrored

Since `pnpm-lock.yaml` doesn't include URLs, there's no need to rewrite that file at all.
___
*[`pkg/ui/BUILD.bazel` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_Aq9v27YlHuwFd9dKN:-N_Aq9v27YlHuwFd9dKO:b-lg4b9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/BUILD.bazel#L29)):*
> ```Starlark
> 
> test_suite(
>     name = "test",
> ```

I was in the neighborhood and realized I never added this.
___
*[`pkg/ui/not-yarn-workspace.sh` line 0 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_AqDf5-gOW_vdML4Hw:-N_AqDf5-gOW_vdML4Hx:bmg51u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/not-yarn-workspace.sh#L0)):*
It's still not a yarn workspace, but now it *is* a pnpm workspace! This script can be completely removed.
___
*[`pkg/ui/package.json` line 329 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_AqMRjEBzOFjER-Zjt:-N_AqMRjEBzOFjER-Zju:bdkvvle) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/package.json#L329)):*
> ```JSON
>       },
> 
>       
> ```

@sjbarag you accidentally some trailing spaces. Get it together, bro.
___
*[`pkg/ui/pnpm-lock.yaml` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_AqXd76lUvjW3w9Vj5:-N_AqXd76lUvjW3w9Vj6:bsrkjfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/pnpm-lock.yaml#L1)):*
> ```YAML
> lockfileVersion: '6.0'
> ```

local engineer finds one weird file to grind browser performance to a halt! JS engines hate him
___
*[`pkg/ui/workspaces/cluster-ui/BUILD.bazel` line 125 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_AqpuNDZW6ZDi8x5qN:-N_AqpuNDZW6ZDi8x5qO:b-ng7chd) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/workspaces/cluster-ui/BUILD.bazel#L125)):*
> ```Starlark
>         # |  _ <| |___  \ V / | |___|  _ < | |   | |  | | |___
>         # |_| \_\_____|  \_/  |_____|_| \_\|_|   |_|  |_|_____|
>         "--bail",
> ```

whoops, right. Thanks `figlet`.



_Code quote:_
```Starlark
        #  ____  _______     _______ ____ _____   __  __ _____
        # |  _ \| ____\ \   / / ____|  _ \_   _| |  \/  | ____|
        # | |_) |  _|  \ \ / /|  _| | |_) || |   | |\/| |  _|
        # |  _ <| |___  \ V / | |___|  _ < | |   | |  | | |___
        # |_| \_\_____|  \_/  |_____|_| \_\|_|   |_|  |_|_____|
        "--bail",
```
___
*[`pkg/ui/workspaces/cluster-ui/jest.config.js` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_AquBL85dxqJr3hahA:-N_AquBL85dxqJr3hahB:bie6eae) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/workspaces/cluster-ui/jest.config.js#L12)):*
> ```JavaScript
> 
> const path = require("path");
> const _ = require("identity-obj-proxy");
> ```

Also not needed
___
*[`pkg/ui/workspaces/db-console/BUILD.bazel` line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_Ar4ok7yxptNlBNscJ:-N_Ar4ok7yxptNlBNscK:b-pht4fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/workspaces/db-console/BUILD.bazel#L62)):*
> ```Starlark
>     outs = [
>         "db-console-ccl/assets/bundle.js",
>         "db-console-ccl/assets/bundle.js.map",
> ```

No longer necessary
___
*[`pkg/ui/workspaces/db-console/BUILD.bazel` line 84 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_Ar71B5JHJm-93T0Av:-N_Ar71B5JHJm-93T0Aw:b-pht4fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/workspaces/db-console/BUILD.bazel#L84)):*
> ```Starlark
>     outs = [
>         "db-console-oss/assets/bundle.js",
>         "db-console-oss/assets/bundle.js.map",
> ```

No longer necessary
___
*[`pkg/ui/workspaces/db-console/webpack.config.js` line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_ArCSC6RJ6IMGxG0dq:-N_ArCSD9dH_DVORL9a2:be940s2) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/workspaces/db-console/webpack.config.js#L88)):*
> ```JavaScript
> 
>     mode: argv.mode || "production",
>     devtool: "source-map",
> ```

Remove this, or we'll ship sourcemaps to users. Which isn't bad from a secrecy perspective. They're just big files, and the `cockroach` binary would get bigger.
___
*[`pkg/ui/workspaces/db-console/src/redux/queryManager/saga.ts` line 183 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_ArVh0BT_o_pPt0_D6:-N_ArVh0BT_o_pPt0_D7:bkxjovw) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/workspaces/db-console/src/redux/queryManager/saga.ts#L183)):*
> ```XML
>       state.sagaTask = (yield fork(managedQuerySaga, state)) as ReturnType<
>         typeof managedQuerySaga
>       >;
> ```

I remember having issues when setting up types for this line. Did this end up working? I assume so, since `./dev ui lint` passes. I may just be making things up now :shrug: 



_Code quote:_
```XML
      state.sagaTask = (yield fork(managedQuerySaga, state)) as ReturnType<
        typeof managedQuerySaga
      >;
```
___
*[`pkg/ui/workspaces/e2e-tests/BUILD.bazel` line 25 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106720#-N_ArhpFAUV8fLDDlp0n:-N_ArhpFAUV8fLDDlp0o:b-c3hbf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc0de01fb8f4d44318ae6808b1bc58d38d3bf7/pkg/ui/workspaces/e2e-tests/BUILD.bazel#L25)):*
> ```Starlark
>             "cypress/e2e/**",
>             "cypress/support/**",
>         ],
> ```

Cypress-produced screenshots and videos have spaces in their filenames, which causes Bazel to break really hard. Not sure how this was working before :shrug: 



_Code quote:_
```Starlark
        include = [
            "cypress/e2e/**",
            "cypress/support/**",
        ],
```

File: pkg/cmd/dev/ui.go [hunk 1]
```diff
@@ -67,6 +69,7 @@ func getUIDirs(d *dev) (*UIDirectories, error) {
 	}
 
 	return &UIDirectories{
+		root:         path.Join(workspace, "./pkg/ui"),
```
Comment (rickystewart) [inline pkg/cmd/dev/ui.go:72]: Nit: this call, and all the other calls to `path.Join` here, should actually be `filepath.Join`. Can you make the update while you're in the area?

File: pkg/cmd/generate-distdir/main.go [hunk 2]
```diff
@@ -163,58 +163,6 @@ func getShasFromNodeRepositoriesCall(call *syntax.CallExpr, shas map[string]stri
 	return nil
 }
 
-func getShasFromYarnRepositoriesCall(call *syntax.CallExpr, shas map[string]string) error {
```
Comment (rickystewart) [inline pkg/cmd/generate-distdir/main.go:166]: This was dead code as of my update to `rules_js`, safe to remove here.

**Review by rickystewart — CHANGES_REQUESTED** [review]
Can you additionally bump the version number in the `dev` top-level script? Will be necessary since we have a backwards-incompatible change.

LGTM once the CI issues are worked through. Thanks for making this change, things are a lot simpler. Having suffered through trying to get the `rules_js` upgrade work in the absence of a workspace tying all the various sub-projects together and keeping the dependencies in sync, I think this will be a big improvement.

OOC are you planning to backport to 23.1?

File: build/teamcity/cockroach/ci/tests/ui_test_impl.sh [hunk 3]
```diff
@@ -3,4 +3,9 @@
 set -xeuo pipefail
 
 bazel build //pkg/cmd/bazci --config=ci
-$(bazel info bazel-bin --config=ci)/pkg/cmd/bazci/bazci_/bazci -- test --config=ci //pkg/ui/workspaces/db-console:jest //pkg/ui/workspaces/cluster-ui:jest
+# Any missing dependencies will cause all of //pkg/ui:test to fail during the
+# dependency-download phase, leading to cryptic failures.
+# Check for unmirrored dependencies before testing the UI to provide a clear
+# error message.
+$(bazel info bazel-bin --config=ci)/pkg/cmd/bazci/bazci_/bazci -- test --config=ci //pkg/cmd/mirror/npm:list_unmirrored_dependencies
```
[the remark this answers is not in the document]
Comment (sjbarag) [reply build/teamcity/cockroach/ci/tests/ui_test_impl.sh:10]: Interesting! That may present weird depending on how it's implemented, but I suppose we'll find out.

**Comment by sjbarag** [comment]
> OOC are you planning to backport to 23.1?

I'm willing to give it a try, but TBH I can't tell how painful it'll be. This was already pretty rough, but I'll poke at it for a few minutes to see.

File: pkg/cmd/dev/ui.go (hunk 1 above)
Comment (sjbarag) [reply pkg/cmd/dev/ui.go:72]: Oh right, good catch. Yep, will do.

**Comment by rickystewart** [comment]
> I'm willing to give it a try, but TBH I can't tell how painful it'll be. This was already pretty rough, but I'll poke at it for a few minutes to see.

No worries, just something I'm wondering about. If we don't backport then it'll have to be factored in as we're doing further build system backports down the line (have to make sure we don't accidentally pull an update to `dev` that is invalid before this change)

**Comment by sjbarag** [comment]
@rickystewart looks like we've got a service-account permissions issue: https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/10890575?buildTab=tests&expandedTest=build%3A%28id%3A10890566%29%2Cid%3A439&showLog=10890566_422_375.380.440.441&logFilter=debug&logView=flowAware

Would you rather we just… not do this step? And let things fail with HTTP 404 responses from cockroach-npm-deps.googleapis.com/foo/-/foo-1.2.3.tgz ? I'm fine with that if you are :)

**Comment by sjbarag** [comment]
Removed the CI-side check for unmirrored dependencies. Should be a bit better now.

**Comment by rickystewart** [comment]
> Would you rather we just… not do this step? And let things fail with HTTP 404 responses from cockroach-npm-deps.googleapis.com/foo/-/foo-1.2.3.tgz ? I'm fine with that if you are :)

I'm fine with this. Presumably devs can put 2 and 2 together in this case.

**Review by rickystewart — APPROVED** [review]
Looks like you just have to deal with the `dev` unit tests and then you're good to go.

**Comment by sjbarag** [comment]
bors r=rickystewart p=1

**Comment by sjbarag** [comment]
bors r=rickystewart p=1

**Pull request merged**